R/write.IAMCxlsx.R

```diff
@@ -0,0 +1,56 @@
+
+#' Write .xlsx file
+#'
+#' A write `.xlsx` file in line with IAMC standard.
+#'
+#' @md
+#' @param x A [`quitte`] data frame.
+#' @param path Path or connection to write to.
+#' @param append Overwrite existing files (`FALSE`, default), or append to them
+#'   (`TRUE`).
+#'
+#' @author Michaja Pehl, Oliver Richters
+#'
+#' @examples
+#' write.IAMCxlsx(quitte_example_data, tempfile())
+#'
+#' @importFrom dplyr mutate rename_with select
+#' @importFrom tidyr pivot_wider
+#' @importFrom tidyselect matches
+#' @importFrom stringr str_to_title
+#' @importFrom writexl write_xlsx
+
+#' @export
+write.IAMCxlsx <- function(x, path, append = FALSE) {
+
+    default_columns <- c('Model', 'Scenario', 'Region', 'Variable', 'Unit',
+                         'Period', 'Value')
+
+    . <- NULL
+
+    # guardians
+    no_quitte <- any(
+        !is.data.frame(x),
+        !all(c('model', 'scenario', 'region', 'variable', 'unit', 'period',
+               'value') %in% tolower(names(x))))
+    if (no_quitte) {
+        stop('x must be a quitte data frame')
+    }
+
+    if (append && file.exists(path)) {
+        x <- rbind(read.quitte(path), x)
+    }
+
+    # make column names upper-case
+    x <- x %>%
+        rename_with(.fn = str_to_title,
+                    .cols = matches('^[A-Za-z][A-Za-z0-9_]*$')) %>%
+        select(setdiff(default_columns, last(default_columns)),
+               setdiff(colnames(.), default_columns),
+               last(default_columns)) %>%
+        arrange(.data$Period) %>%
+        mutate(!!sym('Value') := ifelse(! is.finite(!!sym('Value')) | !!sym('Value') == "", 0, !!sym('Value'))) %>%
```

Comment:
1. Is that desired behaviour?  Seems to me putting/leaving `NA` (or "N/A" or "n_a" or whatever IIASA is using) instead of valid-looking zeros would make more sense …
2. ```
   mutate(Value = ifelse(!is.finite(.data$Value) | .data$Value == "", 0, 
                         .data$Value))
   ```
   is more concise.

Reply:
Asked the IIASA admins about that. I just adopted Jeromes solution for NGFS, but maybe this is not accurate (anymore?).

Reply:
In any case, an easier (and probably faster, but who is counting) way would be
```diff
-        arrange(.data$Period) %>%
-        mutate(!!sym('Value') := ifelse(! is.finite(!!sym('Value')) | !!sym('Value') == "", 0, !!sym('Value'))) %>%
-        pivot_wider(names_from = 'Period', values_from = 'Value')
+        filter(is.finite(.data$Value), '' != .data$Value) %>% 
+        pivot_wider(names_from = 'Period', names_sort = TRUE,
+                    values_from = 'Value', values_fill = 0)
```
If IIASA says "use `'n_a'`", you will have to convert to strings first.